### ISSUE TITLE
New Check: WhitespaceBeforeEmptyBody

### ISSUE THREAD

**Issue by mohitsatr** [checkstyle/checkstyle#17715]
continuation of: #10834 

WhitespaceAround has a lot of properties allowEmptyXxxxxx (they has 'true' in google_checks.xml) that skip violating on code like:
```
/** some. */
public class Test {
  void returnTrue(){}
}
```
Google config:
https://github.com/checkstyle/checkstyle/blob/488018b278b33b5be6ac739846b5960b709d5ab0/src/main/resources/google_checks.xml#L165-L172

---------

WhitespaceBeforeEmptyBody Check should check that a space is present before Empty body of a class, method, interface, static,  constructor , loop(while, do, for) and record.

https://google.github.io/styleguide/javaguide.html#s4.6.2-horizontal-whitespace

one whitespace is required:
> Before any open curly brace ({), with two exceptions:
    @SomeAnnotation({a, b}) (no space is used)
    String[][] x = {{"foo"}}; (no space is required between {{, by item 9 below)

Some of the cases are already covered:

```java
/** some javadoc. */
public class Test {

 
  Test(){}  // ok

  /** some javadoc. */
  public static void main(String... args) {

    boolean b = System.currentTimeMillis() < 0;

    do{} while (b);  // ok 
  
    Runnable noop = () ->{};  // ok
  }


  record Record2(String str) {
    public Record2{} // ok
  }
}
```

```
$ java -jar checkstyle-11.0.0-all.jar -c google_checks.xml Test.java 
Starting audit...
[WARN] /mnt/5D92528E6B945467/test/testing/Test.java:7:30: WhitespaceAround: '{' is not preceded with whitespace. [WhitespaceAround]
[WARN] /mnt/5D92528E6B945467/test/testing/Test.java:14:5: 'do' is not followed by whitespace. [WhitespaceAfter]
[WARN] /mnt/5D92528E6B945467/test/testing/Test.java:14:5: WhitespaceAround: 'do' is not followed by whitespace. Empty blocks                may only be represented as {} when not part of a multi-block statement (4.1.3) [WhitespaceAround]
[WARN] /mnt/5D92528E6B945467/test/testing/Test.java:19:24: '->' is not followed by whitespace. [WhitespaceAfter]
[WARN] /mnt/5D92528E6B945467/test/testing/Test.java:19:24: WhitespaceAround: '->' is not followed by whitespace. Empty blocks                may only be represented as {} when not part of a multi-block statement (4.1.3) [WhitespaceAround]
[WARN] /mnt/5D92528E6B945467/test/testing/Test.java:28:19: WhitespaceAround: '{' is not preceded with whitespace. [WhitespaceAround]
Audit done.
```

Cases which are not covered:

```java
/** some javadoc. */
public class Test {

  Test instance = new Test(){}; // Expected Violation


  void method(){} // Expected Violation

  class Class{} // Expected Violation

  interface Interface{} // Expected Violation

  @interface Annotation{} // Expected Violation

  enum Enum{} // Expected Violation

  record Record(){} // Expected Violation

  /** some javadoc. */
  public static void main(String... args) {

    boolean b = System.currentTimeMillis() < 0;

    while (b){} // Expected Violation

    for (int i = 1; i > 1; i++){} // Expected Violation
  }

  static{} // Expected Violation
}
```

```
$ java -jar checkstyle-11.0.0-all.jar -c google_checks.xml Test.java 
Starting audit...
Audit done.
```

WhitespaceBeforeEmptyBody Check will cover all the above cases, even the ones we already cover.

Example Config of WhitespaceBeforeEmptyBody: 

```xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <module name="TreeWalker">
        <module name="WhitespaceBeforeEmptyBody"/>
    </module>
</module>
```

Example Code:
```java
public class WhitespaceBeforeEmptyBody {

  void method(){}

  class Class{}

```

Example violations:
```
$ java -jar checkstyle-11.0.0-all.jar -c google_checks.xml WhitespaceBeforeEmptyBody.java 
Starting audit...
[WARN] /mnt/5D92528E6B945467/test/testing/WhitespaceBeforeEmptyBody.java:3:15: Whitespace is not present before the empty body of 'method'.[WhitespaceBeforeEmptyBody]
[WARN] /mnt/5D92528E6B945467/test/testing/WhitespaceBeforeEmptyBody.java:5:13: Whitespace is not present before the empty body of 'Class'.[WhitespaceBeforeEmptyBody]
Audit done.
```

**ATTENTION**: update google coverage page to change color of coverage and remove note on partial support. 2 notes.

**Comment by vivek-0509** [comment]
i am on it

**Comment by vivek-0509** [comment]
Hi @romani ,

I think we should also include this additional tokens in WhitespaceBeforeEmptyBodyCheck
` INSTANCE_INIT, LITERAL_SWITCH, LITERAL_TRY, LITERAL_CATCH, LITERAL_FINALLY, and LITERAL_SYNCHRONIZED`.

The Google Style Guide requires a space `"before any open curly brace ({)" with only two exceptions (array and annotation initializers)`, meaning this rule should apply universally to all code block constructs. Importantly, Google Java Formatter automatically adds a space before the { for all these constructs, code formatted with `try{} `becomes `try {},` `switch(x){} `becomes `switch(x) {}` (right curly goes on next line just showed like this for our case), and so on. This means our check should align with the formatter's behavior to maintain consistency.

Should i include this tokens also ??

**Comment by romani** [comment]
>Should i include this tokens also ??

please use google formatter to check if that is required, if formatter is doing whitespace before, we should also do this.

**Comment by vivek-0509** [comment]
I actually verified that  through Google Java Formatter it also enforces a space before {
I have raised the PR please review it https://github.com/checkstyle/checkstyle/pull/18763

here is the google formatter result

Before google formatter:
```
class TestGoogleFormatter {

    private String name;{}

    void testSwitch(int x) {
        switch (x){}
    }

    void testTryCatchFinally() {
        try{}
        catch (Exception e){}
        finally{}
    }

    void testSynchronized() {
        synchronized (this){}
    }
}
```
After google formatter:
```
class TestGoogleFormatter {

  private String name;

  {
  }

  void testSwitch(int x) {
    switch (x) {
    }
  }

  void testTryCatchFinally() {
    try {
    } catch (Exception e) {
    } finally {
    }
  }

  void testSynchronized() {
    synchronized (this) {
    }
  }
}
```

I included INSTANCE_INIT as a default token, though I acknowledge it is an unusual case. Instance initializer blocks typically start on a new line, so violations are rare. However, I kept it in the token list to detect edge cases where someone might accidentally misplace {} on the same line as another statement.Google Formatter handles this by separating it onto its own line

**Comment by romani** [comment]
Thanks a lot, we just need to care about space presence. Location of `}` should be responsibility of other Check.

**Comment by vivek-0509** [comment]
@romani sure will remove INSTANCE_INIT

**Comment by romani** [comment]
fix is merged

**Comment by romani** [comment]
https://github.com/checkstyle/checkstyle/actions/runs/29877487509/job/88791092920

needs to stay open as there is refeence in code for issue

**Comment by romani** [comment]
fix for enablement in google style is merged
